editor/components/inserter/menu.js

```diff
@@ -12,10 +12,161 @@ import Dashicon from 'components/dashicon';
 class InserterMenu extends wp.element.Component {
 	constructor() {
 		super( ...arguments );
+		this.blockTypes = wp.blocks.getBlocks();
+		this.categories = wp.blocks.getCategories();
+		this.nodes = {};
 		this.state = {
-			filterValue: ''
+			filterValue: '',
+			focusedElementRef: null
 		};
 		this.filter = this.filter.bind( this );
+		this.isShownBlock = this.isShownBlock.bind( this );
+		this.changeMenuSelection = this.changeMenuSelection.bind( this );
+		this.setSearchFocus = this.setSearchFocus.bind( this );
+		this.bindReferenceNode = this.bindReferenceNode.bind( this );
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.listVisibleBlocksByCategory = this.listVisibleBlocksByCategory.bind( this );
+	}
+
+	componentDidMount() {
+		document.addEventListener( 'keydown', this.onKeyDown );
+	}
+
+	componentWillUnmount() {
+		document.removeEventListener( 'keydown', this.onKeyDown );
+	}
+
+	isShown( filterValue, string ) {
+		return string.toLowerCase().indexOf( filterValue.toLowerCase() ) !== -1;
+	}
+
+	isShownBlock( filterValue, block ) {
+		return this.isShown( filterValue, block.title );
+	}
+
+	visibleFocusables( nodes, filterValue ) {
+		const focusables = [];
+
+		for ( const refName in nodes ) {
+			if ( this.isShown( filterValue, refName ) && 'search' !== refName ) {
+				focusables.push( refName );
+			}
+		}
+
+		// Add search regardless at the end.
+		focusables.push( 'search' );
```

Comment:
If you take a look at the Slack "inserter" or the MacOS emoji keyboard which are the inspiration for this feature @jasmussen you'll notice the `search` input is always focused when we navigate using the arrows.

But Event in the case we make the search input focusable I'd base my logic on the data instead of the dom nodes:

```js
const visibleBlocks = this.getVisibleBlocks();
const blockToSelect = this.getNextBlock( currentBlock, visibleBlocks );
if ( blockToSelect ) {
  const blockNode = this.nodes[ blockToSelect.blockType ];
  blockNode.focus();
} else {
  this.searchInput.focus():
}
```